src/core/net/srp_server.cpp

```diff
@@ -0,0 +1,1654 @@
+
+static Dns::UpdateHeader::Response ErrorToDnsResponseCode(otError aError)
+{
+    Dns::UpdateHeader::Response responseCode;
+
+    switch (aError)
+    {
+    case OT_ERROR_NONE:
+        responseCode = Dns::UpdateHeader::kResponseSuccess;
+        break;
+    case OT_ERROR_NO_BUFS:
+        responseCode = Dns::UpdateHeader::kResponseServerFailure;
+        break;
+    case OT_ERROR_PARSE:
+        responseCode = Dns::UpdateHeader::kResponseFormatError;
+        break;
+    case OT_ERROR_DUPLICATED:
+        responseCode = Dns::UpdateHeader::kResponseNameExists;
+        break;
+    default:
+        responseCode = Dns::UpdateHeader::kResponseRefused;
+        break;
+    }
+
+    return responseCode;
+}
+
+Server::Server(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mSocket(aInstance)
+    , mAdvertisingHandler(nullptr)
+    , mAdvertisingHandlerContext(nullptr)
+    , mDomain(nullptr)
+    , mMinLease(kDefaultMinLease)
+    , mMaxLease(kDefaultMaxLease)
+    , mMinKeyLease(kDefaultMinKeyLease)
+    , mMaxKeyLease(kDefaultMaxKeyLease)
+    , mLeaseTimer(aInstance, HandleLeaseTimer, this)
+    , mOutstandingUpdatesTimer(aInstance, HandleOutstandingUpdatesTimer, this)
+    , mEnabled(false)
+{
+    IgnoreError(SetDomain(kDefaultDomain));
+}
+
+Server::~Server(void)
+{
+    GetInstance().HeapFree(mDomain);
+}
+
+void Server::SetServiceHandler(otSrpServerServiceUpdateHandler aServiceHandler, void *aServiceHandlerContext)
+{
+    mAdvertisingHandler        = aServiceHandler;
+    mAdvertisingHandlerContext = aServiceHandlerContext;
+}
+
+bool Server::IsRunning(void) const
+{
+    return mSocket.IsBound();
+}
+
+void Server::SetEnabled(bool aEnabled)
+{
+    VerifyOrExit(mEnabled != aEnabled);
+
+    mEnabled = aEnabled;
+
+    if (!mEnabled)
+    {
+        Stop();
+    }
+    else if (Get<Mle::MleRouter>().IsAttached())
+    {
+        Start();
+    }
+
+exit:
+    return;
+}
+
+otError Server::SetLeaseRange(uint32_t aMinLease, uint32_t aMaxLease, uint32_t aMinKeyLease, uint32_t aMaxKeyLease)
+{
+    otError error = OT_ERROR_NONE;
+
+    // TODO: Support longer LEASE.
+    // We use milliseconds timer for LEASE & KEY-LEASE, this is to avoid overflow.
+    VerifyOrExit(aMaxKeyLease <= TimerMilli::kMaxDelay / 1000, error = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(aMinLease <= aMaxLease, error = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(aMinKeyLease <= aMaxKeyLease, error = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(aMinLease <= aMinKeyLease, error = OT_ERROR_INVALID_ARGS);
+    VerifyOrExit(aMaxLease <= aMaxKeyLease, error = OT_ERROR_INVALID_ARGS);
+
+    mMinLease    = aMinLease;
+    mMaxLease    = aMaxLease;
+    mMinKeyLease = aMinKeyLease;
+    mMaxKeyLease = aMaxKeyLease;
+
+exit:
+    return error;
+}
+
+uint32_t Server::GrantLease(uint32_t aLease) const
+{
+    OT_ASSERT(mMinLease <= mMaxLease);
+
+    return (aLease == 0) ? 0 : OT_MAX(mMinLease, OT_MIN(mMaxLease, aLease));
```

Comment:
What if `SetLeaseRange()` is called after `GrantLease()`?

Reply:
Granted lease will not be affected, and new service will be granted with new lease range.